File: src/main/java/org/radarbase/management/security/DomainUserDetailsService.java
package org.radarbase.management.security;

import org.radarbase.management.domain.User;
import org.radarbase.management.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

/**
 * Authenticate a user from the database.
 */
@Component("userDetailsService")
public class DomainUserDetailsService implements UserDetailsService {

    private static final Logger log = LoggerFactory.getLogger(DomainUserDetailsService.class);

    private final UserRepository userRepository;

    public DomainUserDetailsService(
            UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    @Transactional
    public UserDetails loadUserByUsername(final String login) {
        log.debug("Authenticating {}", login);
        String lowercaseLogin = login.toLowerCase(Locale.ENGLISH);
        User user = userRepository.findOneWithRolesByLogin(lowercaseLogin)
                .orElseThrow(() -> new UsernameNotFoundException(
                        "User " + lowercaseLogin + " was not found in the database"));
        if (!user.activated) {
            throw new UserNotActivatedException("User " + lowercaseLogin
                    + " was not activated");
        }

        List<GrantedAuthority> grantedAuthorities = user.getAuthorities().stream()
                .map(authority -> new SimpleGrantedAuthority(authority.name))
                .collect(Collectors.toList());

        return new org.springframework.security.core.userdetails.User(
                lowercaseLogin,
                user.password,
                grantedAuthorities);
    }
}

File: src/main/java/org/radarbase/management/security/DomainUserDetailsService.kt
package org.radarbase.management.security

import org.radarbase.management.repository.UserRepository
import org.slf4j.LoggerFactory
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

/**
 * Authenticate a user from the database.
 */
@Component("userDetailsService")
open class DomainUserDetailsService(
    private val userRepository: UserRepository
) : UserDetailsService {
    @Transactional
    override fun loadUserByUsername(login: String): UserDetails {
        log.debug("Authenticating {}", login)
        val lowercaseLogin = login.lowercase()
        val user = userRepository.findOneWithRolesByLogin(lowercaseLogin)
            ?: throw UsernameNotFoundException(
                    "User $lowercaseLogin was not found in the database"
                )
        if (!user.activated) {
            throw UserNotActivatedException(
                "User " + lowercaseLogin
                        + " was not activated"
            )
        }
        val grantedAuthorities =
            user.authorities!!.map { authority -> SimpleGrantedAuthority(authority) }
        return User(
            lowercaseLogin,
            user.password,
            grantedAuthorities
        )
    }

    companion object {
        private val log = LoggerFactory.getLogger(DomainUserDetailsService::class.java)
    }
}
